src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -726,34 +751,60 @@ async_manifest_view::get_term_last_offset(model::term_id term) noexcept {
 
         auto res = co_await get_cursor(cursor_start);
         if (res.has_error()) {
-            vlog(_ctxlog.error, "Failed to scan metadata: {}", res.error());
+            _ctxlog.log(
+              log_level_for_error(res.error()),
+              "Failed to scan metadata: {}",
+              res.error());
             co_return res.as_failure();
         }
-        std::optional<kafka::offset> res_offset;
-        co_await ss::repeat(
-          [this, &res_offset, term, cursor = std::move(res.value())] {
-              const auto& manifest = *cursor->manifest();
-              vlog(
-                _ctxlog.debug,
-                "Scanning manifest {} for term {}",
-                manifest.get_manifest_path(),
-                term);
-              for (auto meta : manifest) {
-                  if (meta.segment_term > term) {
-                      res_offset = meta.base_kafka_offset() - kafka::offset(1);
-                      vlog(
-                        _ctxlog.debug,
-                        "Scan found offset {} at term {}",
-                        res_offset.value(),
-                        meta.segment_term);
-                      return ss::make_ready_future<ss::stop_iteration>(
-                        ss::stop_iteration::yes);
-                  }
-              }
-              return cursor->next_iter();
-          });
 
-        co_return res_offset;
+        try {
+            std::optional<kafka::offset> res_offset;
+            co_await ss::repeat(
+              [this, &res_offset, term, cursor = std::move(res.value())] {
+                  const auto& manifest = *cursor->manifest();
+                  vlog(
+                    _ctxlog.debug,
+                    "Scanning manifest {} for term {}",
+                    manifest.get_manifest_path(),
+                    term);
+                  for (auto meta : manifest) {
+                      if (meta.segment_term > term) {
+                          res_offset = meta.base_kafka_offset()
+                                       - kafka::offset(1);
+                          vlog(
+                            _ctxlog.debug,
+                            "Scan found offset {} at term {}",
+                            res_offset.value(),
+                            meta.segment_term);
+                          return ss::make_ready_future<ss::stop_iteration>(
+                            ss::stop_iteration::yes);
+                      }
+                  }
+                  return cursor->next_iter();
+              });
+
+            co_return res_offset;
+        } catch (const std::system_error& e) {
+            // thrown by `async_manifest_view::next_iter`
+            const auto err_outcome = static_cast<error_outcome>(
```

Comment:
I think that the correct way to do this is to make a temporary error code and compare with it:

```
auto shutdown_errc = make_error_code(error_outcome::shutting_down);
if (e.code() == shutdown_errc) { ... use debug level for logging ... }
```

Reply:
I agree with the other two instance you highlighted, but here we want to convert from a `std::error_code` to a `cloud_storage::error_outcome`. Looking at the def of `make_error_code`, looks like `static_cast` is the way to go.

Reply:
Should we at least check `e.code().category()` before casting? Or are `system_errors` from this code snippet not possible?

Reply:
Shouldn't be the case, but we can check

Reply:
Added a category check.